gossip/src/cluster_info.rs

```diff
@@ -208,6 +207,73 @@ fn should_retain_crds_value(
     }
 }
 
+/// A sender implementation that evicts the oldest message when the channel is full.
+#[derive(Clone)]
+pub(crate) struct EvictingSender<T> {
+    sender: Sender<T>,
+    receiver: Receiver<T>,
+}
+
+impl<T> EvictingSender<T> {
+    /// Create a new evicting sender with provided sender, receiver.
+    #[inline]
+    pub(crate) fn new(sender: Sender<T>, receiver: Receiver<T>) -> Self {
+        Self { sender, receiver }
+    }
+
+    /// Create a new `EvictingSender` with a bounded channel of the specified capacity.
+    #[inline]
+    pub(crate) fn new_bounded(capacity: usize) -> (Self, Receiver<T>) {
+        let (sender, receiver) = bounded(capacity);
+        (Self::new(sender, receiver.clone()), receiver)
+    }
+}
+
+impl<T> ChannelSend<T> for EvictingSender<T>
+where
+    T: Send + 'static,
+{
+    #[inline]
+    fn send(&self, msg: T) -> std::result::Result<(), SendError<T>> {
+        self.sender.send(msg)
+    }
+
+    fn try_send(&self, msg: T) -> std::result::Result<(), TrySendError<T>> {
+        let Err(e) = self.sender.try_send(msg) else {
+            return Ok(());
+        };
+
+        match e {
+            // Prefer newer messages over older messages.
+            TrySendError::Full(msg) => match self.receiver.try_recv() {
+                Ok(older) => {
+                    // Attempt to to requeue the newer message.
```

Comment:
nit: "to to" typo, and I'd add a comment here that says that there's a race condition if you have multiple senders, something like 

```
Attempt to requeue the newer message. NB: if multiple senders are used and another sender is faster than us to send() after we've popped `older`,  our try_send() will fail with Full(msg), in which case we drop the new message. 
```